libraries/AP_ToshibaCAN/AP_ToshibaCAN.cpp

```diff
@@ -0,0 +1,287 @@
+static const uint8_t COMMAND_MOTOR3 = 0x3B;
+static const uint8_t COMMAND_MOTOR2 = 0x3D;
+static const uint8_t COMMAND_MOTOR1 = 0x3F;
+
+// data format for messages from ESC
+static const uint8_t MOTOR_DATA1 = 0x40;
+static const uint8_t MOTOR_DATA2 = 0x50;
+static const uint8_t MOTOR_DATA3 = 0x60;
+static const uint8_t MOTOR_DATA5 = 0x80;
+
+// processing definitions
+static const uint16_t TOSHIBACAN_OUTPUT_MIN = 6300;
+static const uint16_t TOSHIBACAN_OUTPUT_MAX = 32000;
+static const uint16_t SET_PWM_MIN_INTERVAL_US = 2500;
+static const uint32_t SET_PWM_TIMEOUT_US = 2000;
+static const uint8_t CAN_IFACE_INDEX = 0;
+
+AP_ToshibaCAN::AP_ToshibaCAN()
+{
+    debug_can(2, "ToshibaCAN: constructed\n\r");
+}
+
+AP_ToshibaCAN *AP_ToshibaCAN::get_tcan(uint8_t driver_index)
+{
+    if (driver_index >= AP::can().get_num_drivers() ||
+        AP::can().get_protocol_type(driver_index) != AP_BoardConfig_CAN::Protocol_Type_ToshibaCAN) {
+        return nullptr;
+    }
+    return static_cast<AP_ToshibaCAN*>(AP::can().get_driver(driver_index));
+}
+
+// initialise ToshibaCAN bus
+void AP_ToshibaCAN::init(uint8_t driver_index)
+{
+    _driver_index = driver_index;
+
+    debug_can(2, "ToshibaCAN: starting init\n\r");
+
+    if (_initialized) {
+        debug_can(1, "ToshibaCAN: already initialized\n\r");
+        return;
+    }
+
+    AP_HAL::CANManager* can_mgr = hal.can_mgr[driver_index];
+
+    if (can_mgr == nullptr) {
+        debug_can(1, "ToshibaCAN: no mgr for this driver\n\r");
+        return;
+    }
+
+    if (!can_mgr->is_initialized()) {
+        debug_can(1, "ToshibaCAN: mgr not initialized\n\r");
+        return;
+    }
+
+    _can_driver = can_mgr->get_driver();
+
+    if (_can_driver == nullptr) {
+        debug_can(1, "ToshibaCAN: no CAN driver\n\r");
+        return;
+    }
+
+    // start calls to loop in separate thread
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_ToshibaCAN::loop, void), _thread_name, 4096, AP_HAL::Scheduler::PRIORITY_CAN, 0)) {
+        debug_can(1, "ToshibaCAN: couldn't create thread\n\r");
+        return;
+    }
+
+    _initialized = true;
+
+    debug_can(2, "ToshibaCAN: init done\n\r");
+
+    return;
+}
+
+// loop to send output to ESCs in background thread
+void AP_ToshibaCAN::loop()
+{
+    uavcan::MonotonicTime timeout;
+    const uint32_t loop_interval_us = MIN(AP::scheduler().get_loop_period_us(), SET_PWM_MIN_INTERVAL_US);
+
+    while (true) {
+        if (!_initialized) {
+            // if not initialised wait 2ms
+            debug_can(2, "ToshibaCAN: not initialized\n\r");
+            hal.scheduler->delay_microseconds(2000);
+            continue;
+        }
+
+        // check for updates
+        if (last_update_us == last_send_us) {
+            hal.scheduler->delay_microseconds(50);
+            continue;
+        }
+
+        _attempts++;
+
+        // take semaphore to read scaled motor outputs
+        if (!_rc_out_sem.take(1)) {
```

Comment:
much better to use a blocking semaphore take, using WITH_SEMAPHORE()
The timeout based semaphore take is much less efficient